platform-service-framework/src/main/java/org/hypertrace/core/serviceframework/DefaultPlatformServiceLifecycle.java

```diff
@@ -0,0 +1,26 @@
+package org.hypertrace.core.serviceframework;
+
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+
+class DefaultPlatformServiceLifecycle implements PlatformServiceLifecycle {
+  private volatile State state = State.NOT_STARTED;
+  private final CompletableFuture<Void> shutdown = new CompletableFuture<>();
+
+  @Override
+  public CompletionStage<Void> shutdownComplete() {
+    return this.shutdown.minimalCompletionStage();
```

Comment:
what is the use of this ?  There used to be abstract doStop() method - wouldn't that work ?

Reply:
doStop does give each service a handle to do things on shutdown, and this could be implemented in each service using that (example: https://github.com/hypertrace/hypertrace-core-graphql/blob/2ad528d40f0c395ed1ba3ea21c88c433766b9862/hypertrace-core-graphql-service/src/main/java/org/hypertrace/core/graphql/service/GraphQlService.java#L72 )

The main issue with that approach is that it requires the same logic or equivalent to be implemented in each service - some way of propagating the shutdown to any listeners. By passing down a lifecycle object (or just the completion stage) we allow this to work neatly with IoC containers so any class can just inject the service lifecycle. An example of using the lifecycle with that pattern can be seen here: https://github.com/hypertrace/hypertrace-core-graphql/blob/main/hypertrace-core-graphql-grpc-utils/src/main/java/org/hypertrace/core/graphql/utils/grpc/DefaultGrpcChannelRegistry.java

The one gotcha right now is that the way we generally separate code concerns is a service module (that is aware of the framework) and an impl module which is agnostic to the specific server container it's running on. Ideally, `PlatformServiceLifecycle` would be moved into an SPI package that impls could rely on without being tied to any specific container, but that felt like overkill at the moment. So we can either make that separation here too, or I was planning on just using fragments of this and injecting the completion stage rather than the full lifecycle.

Reply:
So I pulled out the SPI project in the latest commit. I can revert that if people disagree though.

Reply:
Here's an example of how this change would be used: https://github.com/hypertrace/query-service/compare/update-attribute-client?expand=1

Reply:
thanks for explanation